jax/_src/lru_cache.py

```diff
@@ -0,0 +1,115 @@
+# Copyright 2024 The JAX Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import annotations
+
+import heapq
+import logging
+import os
+import pathlib
+from typing import Any
+
+from jax._src import config
+# from jax._src import path as pathlib
+from jax._src.compilation_cache_interface import CacheInterface
+from jax._src.gfile_cache import GFileCacheImpl
+
+
+filelock: Any = None
+try:
+  import filelock
+except ImportError:
+  pass
+
+
+logger = logging.getLogger(__name__)
+
+
+class Impl:
+
+  def __init__(self, path: str, *, max_size_bytes: int, timeout_secs: float | None=10):
+    path = pathlib.Path(path)
+
+    if max_size_bytes != -1 and filelock is None:
+      raise RuntimeError("Please install filelock package to set `jax_compilation_cache_max_size`")
+
+    self.path = path
+    self.max_size_bytes = max_size_bytes
+    path.mkdir(parents=True, exist_ok=True)
+    self.timeout_secs = timeout_secs
+
+    self.lock_file = self.path / ".lock_path"
+    self.lock = filelock.FileLock(self.lock_file)
+
+  def get(self, key: str) -> bytes | None:
+    file_path = self.path / key
+
+    with self.lock.acquire(timeout=self.timeout_secs):
+      if not os.path.exists(file_path):
+        logger.debug(f"Cache miss for key {key}")
+        return None
+
+      logger.debug(f"Cache hit for key {key}")
+      file_path.touch()  # update mtime
+      return file_path.read_bytes()
+
+  def put(self, key: str, val: bytes) -> None:
+    file_path = self.path / key
+
+    with self.lock.acquire(timeout=self.timeout_secs):
+      if file_path.exists():
+        return
+
+      self._evict_if_needed()
+      file_path.write_bytes(val)
+
+  def _evict_if_needed(self) -> None:
+    if self.max_size_bytes == -1:
+      return  # no eviction if `max_size_bytes` is set to -1
+
+    mtime_path_sizes_heap = []
```

Comment:
How about just `heap` or even `h`. You can add a one-line comment explaining that the heap uses `mtime` as the key.

Reply:
Done

Reply:
Done